aws-rds-cfn-common/src/main/java/software/amazon/rds/common/handler/Tagging.java

```diff
@@ -19,19 +24,83 @@
 import software.amazon.awssdk.services.rds.model.Tag;
 import software.amazon.cloudformation.proxy.ProgressEvent;
 import software.amazon.cloudformation.proxy.ProxyClient;
+import software.amazon.rds.common.error.ErrorCode;
 import software.amazon.rds.common.error.ErrorRuleSet;
+import software.amazon.rds.common.error.ErrorStatus;
+import software.amazon.rds.common.error.NoOpErrorRuleSet;
+import software.amazon.rds.common.error.UnexpectedErrorStatus;
 
 public final class Tagging {
 
+    @Builder(toBuilder = true)
+    @AllArgsConstructor
+    @Data
+    public static class TagSet {
+        @Builder.Default
+        private Set<Tag> systemTags = new HashSet<>();
+        @Builder.Default
+        private Set<Tag> stackTags = new HashSet<>();
+        @Builder.Default
+        private Set<Tag> resourceTags = new HashSet<>();
+
+        public boolean isEmpty() {
+            return systemTags.isEmpty() &&
+                    stackTags.isEmpty() &&
+                    resourceTags.isEmpty();
+        }
+
+        public static TagSet emptySet() {
+            return TagSet.builder().build();
+        }
+    }
+
+    static class SoftFailErrorRuleSet implements ErrorRuleSet {
+        @Override
+        public ErrorStatus handle(Exception exception) {
+            if (exception instanceof AwsServiceException) {
+                final ErrorCode errorCode = ErrorCode.fromException((AwsServiceException) exception);
+                if (ErrorCode.AccessDenied.equals(errorCode)) {
+                    return ErrorStatus.ignore();
+                }
+            }
+            return new UnexpectedErrorStatus(exception);
+        }
+    }
+
+    public static TagSet exclude(final TagSet from, final TagSet what) {
+        final Set<Tag> systemTags = new HashSet<>(from.getSystemTags());
+        systemTags.removeAll(what.getSystemTags());
+
+        final Set<Tag> stackTags = new HashSet<>(from.getStackTags());
+        stackTags.removeAll(what.getStackTags());
+
+        final Set<Tag> resourceTags = new HashSet<>(from.getResourceTags());
+        resourceTags.removeAll(what.getResourceTags());
+
+        return TagSet.builder()
+                .systemTags(systemTags)
+                .stackTags(stackTags)
+                .resourceTags(resourceTags)
+                .build();
+    }
+
     public static <K, V> Map<K, V> mergeTags(Map<K, V> tagsMap1, Map<K, V> tagsMap2) {
         final Map<K, V> result = new HashMap<>();
         result.putAll(Optional.ofNullable(tagsMap1).orElse(Collections.emptyMap()));
         result.putAll(Optional.ofNullable(tagsMap2).orElse(Collections.emptyMap()));
         return result;
     }
 
-    public static Set<Tag> translateTagsToSdk(Map<String, String> tags) {
-        return tags.entrySet()
+    public static Set<Tag> translateTagsToSdk(final TagSet tagSet) {
+        final Set<Tag> allTags = new HashSet<>();
+        allTags.addAll(tagSet.getSystemTags());
```

Comment:
Changed the approach: `CreateHandler` promotes a 2-stage tag create process. Hope it makes more sense.

Reply:
System tags can also be added during updates in case of a resource import update. The handler request will contain system tag values.

I think this should be handled though with the way updateTags is implemented but wouldn't hurt to update any resource import related integ tests to verify.